### DISCUSSION TITLE
Transform into full clock app

### DISCUSSION THREAD

**Discussion by qw123wh** [WrichikBasu/ShakeAlarmClock#36] [Feature requests]
Hi the thing I like most about your app is the ability to turn off the alarm with the power button.  It would be great if this app becomes a complete app in the future, with stopwatch, timer, clock and alarm.  Could you do it?  Perhaps also adding power off alarm, so that the alarm sounds even when the phone is off.

**Comment by WrichikBasu** [comment]
Thank you for your interest in this app!

> It would be great if this app becomes a complete app in the future, with stopwatch, timer, clock and alarm. Could you do it?

In principle, yes. But it will be a long project where I will have to completely rebuild the main activity (that shows the list of alarms). My first priority is to deal with some important bugs. Once that is done, I will consider your suggestion.

> Perhaps also adding power off alarm, so that the alarm sounds even when the phone is off.

You mean when the device is switched off? Unfortunately, that is not possible, because when the device switches off, the OS itself shuts down, so no software can run on it. In fact, no app can work when the device is switched off, even with root permissions.

**Comment by qw123wh** [comment]
Yes i mean when the phone is off.  For phones with snapdragon there is power off alarm, with the following permission org.codeaurora.permission.POWER_OFF_ALARM.  
And no root is needed.  This is possible on aosp clock, like this
https://github.com/Havoc-OS/android_packages_apps_DeskClock

If you extract the apk of this rom and install it on any phone with snapdragon without root (except xiaomi, Miui rom's fault) the alarm will work when the phone is off.  If you want, I'll send you the apk if you have a snapdragon phone.  It is a feature of the snapdragon processor

![3F9388C0-E41A-4A50-A0FF-E1633A44D4DA](https://user-images.githubusercontent.com/59417474/103467339-7f180300-4d4e-11eb-8287-c963ec11f10b.jpeg)
![0F9BE714-F768-4A1E-B3D6-A28565486040](https://user-images.githubusercontent.com/59417474/103467340-80493000-4d4e-11eb-88f3-be04d85a23f1.jpeg)
![9BF3FA65-43DE-4B05-BD05-4BF24A08164D](https://user-images.githubusercontent.com/59417474/103467342-80e1c680-4d4e-11eb-9d40-6a7a538c0be3.jpeg)

**Reply by WrichikBasu** [reply]
Interesting. I am hearing for the first time that alarms can go off even if the phone is switched off. I was searching online, and found that OnePlus devices do have this feature. Even if all snapdragon processors have this feature, it would have to be enabled in the firmware, because from Android, I cannot directly communicate with the processor.

Sometimes, stock apps have elevated permissions which allow them to do some extra functions which third-party apps cannot do. For example, on Samsung Galaxy On7, whenever an alarm is set, an alarm clock icon pops up on the status bar. In Samsung Galaxy M31s, the icon will pop up only if the alarm is set using the stock app.

Anyway, it is certainly possible to add a feature that will work on only some devices, but I would have to learn first how this works. No documentation seems to exist on this topic. I would love to integrate this feature in my app, but will have to do some searching.

I do have a snapdragon phone, but it's Samsung, and the stock alarm app doesn't have this feature. If possible, please upload the apk and I will test it on my device.

**Comment by qw123wh** [comment]
Ok I'll send you the apk, however, make sure you give permission as in the video.  Then set the alarm after at least 15 minutes.  Then turn off the phone, and the phone should turn on again 1 minute before you set the alarm.  Then after the phone stays on. This apk was extracted from a pie rom.

https://drive.google.com/file/d/1QxWf4Zzmd-f0salDvpgDk1tRcb-Xp47N/view?usp=drivesdk

let me know if it works

https://github.com/Havoc-OS/android_packages_apps_DeskClock/search?q=power+off+alarm

https://user-images.githubusercontent.com/59417474/103475718-17929f80-4db0-11eb-8c42-d7b7260cdc6d.MP4

**Comment by WrichikBasu** [comment]
Thanks for the apk. I installed it on my device, and the moment I opened it, the app crashed. I believe my device does not support power off alarms even though it has the snapdragon processor — probably prohibited by the firmware.


> https://github.com/Havoc-OS/android_packages_apps_DeskClock/search?q=power+off+alarm

This link was particularly helpful in finding the exact location of the code. That ROM is sending a broadcast to a qualcomm package to set the power off alarm:

```java
    private static final String ACTION_SET_POWEROFF_ALARM =
            "org.codeaurora.poweroffalarm.action.SET_ALARM";


    private static final String ACTION_CANCEL_POWEROFF_ALARM =
            "org.codeaurora.poweroffalarm.action.CANCEL_ALARM";


    private static final String POWER_OFF_ALARM_PACKAGE =
            "com.qualcomm.qti.poweroffalarm";


    private static final String TIME = "time";


...

    private static void setPowerOffAlarm(Context context, AlarmInstance instance) {
         LogUtils.i("Set next power off alarm : instance id "+ instance.mId);
         Intent intent = new Intent(ACTION_SET_POWEROFF_ALARM);
         intent.addFlags(Intent.FLAG_RECEIVER_FOREGROUND);
         intent.setPackage(POWER_OFF_ALARM_PACKAGE);
         intent.putExtra(TIME, instance.getAlarmTime().getTimeInMillis());
         context.sendBroadcast(intent);
    }
```

It's pretty easy to implement this, but I need to be sure that ROMs from different manufacturers (that support power off alarm) are doing it the same way.

**Reply by qw123wh** [reply]
Maybe your android version is very old on your samsung with snapdragon, I tried it on oppo android 9 with snapdragon and it works fine. Now that you make me think about crash on huawei with android 7. So i think it takes at least android 8-9 i don't know.
Try it on other non snapdragon phones too with min android 8-9, power off alarm will definitely not work, but it shouldn't crash.
Check these two commits as well.  They will surely help you

https://github.com/Havoc-OS/android_packages_apps_DeskClock/commit/42446e490f512d3de00d11fef2b0e9f04b98a0df

https://github.com/Havoc-OS/android_packages_apps_DeskClock/commit/71e534338fde712117eebaa7d36c403559e3e1c2

**Reply by qw123wh** [reply]
if you can implement it you can send me a test apk and I will let you know if it works on my device

**Reply by WrichikBasu** [reply]
Well, then I will implement it and upload the test apk. Note that if you install that one, and later downgrade to the stable version, you might lose the alarms that you had added previously.

**Reply by qw123wh** [reply]
Ok no problem. Thanks

**Reply by WrichikBasu** [reply]
@qw123wh I have implemented the code, please find the apk from [Google Drive](https://drive.google.com/file/d/1Xra3k1WbR2N8BVEOclJF_EStvIb_F64s/view?usp=sharing). I have done some initial testing on my Samsung devices (which, of course, don't have that feature), and there were no crashes. Let me know whether it works.

**Comment by qw123wh** [comment]
I uninstalled the app, then I installed this test apk.  
Then the phone turns on correctly, but the alarm does not ring, but the app crashes, I'll send you the log

https://del.dog/crudozimit


You're on the right track, I'm sure you can make it work.

**Reply by WrichikBasu** [reply]
Okay, it's good news that the phone turns on correctly. I have corrected the exception. What happens when you try this [new apk](https://drive.google.com/file/d/1AiHvb37qHC4Ngg7tNcm8kWWCjIVyKkxe/view?usp=sharing)?

**Reply by qw123wh** [reply]
now i try this again

**Comment by qw123wh** [comment]
however, the alarm does not work even with the phone on, here for you another log

https://del.dog/domylittyr

**Reply by qw123wh** [reply]
this is the previous one

**Reply by WrichikBasu** [reply]
Hmm, that's not supposed to happen. Can you install the stable version and confirm that the alarm is working normally as it should?

**Reply by qw123wh** [reply]
I tried the second apk and it doesn't ring when the phone is on, this is the log.         https://del.dog/stemepathu

This other log is with the phone off

https://del.dog/ughilyghep.txt

I tried the stable app from the play store but that doesn't work either.
This is the log of the stable version of the play store.

https://del.dog/imulivirra.txt

practically always the same error in all three versions (the two test versions and the official one of the play store).

Anyway if it can help you I have android 10 AOSP (custom rom).
Try to apply power off alarm to the code of the old version of the app and see if it works

**Reply by WrichikBasu** [reply]
If you get that error on the stable version too, then I won't be able to sleep tonight, because tomorrow morning, my Google Play console will flood with crash reports. I never asked for phone state permission. Why should that throw an exception?

You will have to wait a bit because I have to first correct the problem in the stable version.

Thanks for your interest, cooperation and patience. Will get back to you soon!

**Reply by qw123wh** [reply]
however, in the stable version prior to 1.3.3 I didn't have any kind of problem, it is something you changed in version 1.3.4 that causes the error on my phone.

**Comment by WrichikBasu** [comment]
I changed some things, but I also tested the changes thoroughly before
publishing the update. The error you have reported is really odd because it
says that the app is denied a permission that I never requested in the
first place.

**Comment by qw123wh** [comment]
try doing a test apk with that permission and see if it works, or there is some other problem at the source.  I know it is a very strange permission to access the phone and also risky, but let's try.  But for safety in the trial apk remove internet access, so at least access to the read privileged phone state is more secure

**Comment by WrichikBasu** [comment]
Till now, I haven't got any crash reports regarding this from Google Play. Keeping fingers crossed.

For the while, unfortunately, I have decided to discontinue work on this power off alarm feature. While I agree that it is not impossible to implement this feature, it is certainly more tiresome to design something that works on only some types of devices/ROMs. Some manufacturers enable this feature in the stock ROM itself, but I spoke to some developers last night, and they said that the permission is reserved only for system apps. Custom ROMs are probably more lenient and allow third-party apps to access the permission, but there are hundreds of custom ROMs out there, and something that works on one may not work on another.

If you still face that error after updating to the latest stable version, try uninstalling the app and rebooting the device, and then install it again.

If someone, in the future, implements this feature and opens a PR, I will be happy to test it and merge it.

**Comment by qw123wh** [comment]
I don't know what happened anyway I performed a reset on the phone I reinstalled the rom, and now that problem gives me even with version 1.3.3 downloaded from github.

I really don't know where the problem lies because now even the official versions don't work for me even the old versions.

**Comment by qw123wh** [comment]
Good news.

The bug has been fixed, I tried this experiment on previous versions app as well, including the two test versions with power off alarm.  Basically the bug is caused by the alarm ringtone, if I do not touch anything and I remain all by default the alarm does not sound, it crashes and always gives me the same error on the log.  If I magically change the alarm ringtone, the alarm rings, the app no ​​longer crashes.  Sounds strange but all the fuss is caused by the ringtone.

However I was convinced that power off alarm, and that the custom rom were not the problem of the bug and I was right. I don't know if you can change something on the ringtone otherwise I will always use this trick and everything will work.

**Reply by WrichikBasu** [reply]
There should not have been a problem if you choose default alarm tone, because I use [`Settings.System.DEFAULT_ALARM_ALERT_URI`](https://developer.android.com/reference/android/provider/Settings.System#DEFAULT_ALARM_ALERT_URI) to get the default tone, and that should never return `null`. In fact, the stack trace that you uploaded at that time pointed out a very weird issue with a permission that I never asked for, and had no relation to this.

Anyway, happy to have you back as a user of my app ;-)

**Comment by qw123wh** [comment]
anyway i found another bug, i have to try it on other phones too, maybe it's always my phone's fault.  Basically, if I set an alarm and then turn off and on again or restart the phone, when the alarm sounds, I can't turn it off with the power button, but only with a shake. And from that moment on I can no longer turn it off with the power button.  I have to uninstall and reinstall the app to make it work.

**Comment by qw123wh** [comment]
As far as power off alarm is concerned, I know that you no longer want to hear about it anyway with the ringer trick I told you earlier and in the test versions I received this other error https://del.dog/ickexethem.txt

Just take a look at it, maybe it can be used for something else.  However, I found the way to make the alarm go off when it was off thanks to you with your app

**Comment by WrichikBasu** [comment]
I tested on all the devices I had, and found that I could turn off the alarm with the power button even after a reboot. Regarding the stack trace you posted above, is that on the stable version? Note that the power-off alarm version that I gave you was in alpha stage, and may not work as intended.

**Reply by qw123wh** [reply]
However I will do some tests on this bug of the power button perhaps with little you can solve.  The power button bug occurs in all versions of the app.  As for the log from the second test apk, this happens with the automatic switching on of the phone (rightly caused by a power off alarm), then when it comes to ringing it crashes and gives me that log, however I think it happens because it was in alpha stage

**Reply by WrichikBasu** [reply]
If you want, you can create an issue for the power button bug (mention stable release rather than alpha release). My tests on Samsung devices did not reveal the bug, but that doesn't mean that it can't exist on devices from other manufacturers.
